man/man1/squeeze.1
.TH squeeze 1
.SH NAME
\fBsqueeze\fR - Finds the best subset, of the specified files, which fits into a given space.
.SH SYNOPSIS
\fBsqueeze\fR [\fIOPTIONS\fR] [\fIFile-path\fR ...]
.SH DESCRIPTION
.PP
Finds the subset of the specified \fIFile-path\fRs, which fits into the specified space with least room to spare, whilst also meeting the minimum usage-requirements;
i.e. a degenerate instance of the "0-1 Knapsack-problem".
.PP
Any directories amongst the specified \fIFile-path\fRs are treated as atomic units, & therefore only solutions which involve either all or none of the files contained, are returned.
.PP
Because of its exponential time-complexity, solutions of increasing suitability are produced lazily,
rather than waiting until the optimal solution is known (which might take an inordinately long time).
.SH OPTIONS
.TP
\fB--verbosity=\fR(\fBSilent\fR|\fBNormal\fR|\fBVerbose\fR|\fBDeafening\fR)
Produces additional explanatory output where appropriate.
.br
CAVEAT: to be effective, this option must precede others.
.SS "Generic Program-information"
.TP
\fB-v\fR, \fB--version\fR
Outputs version-information, & then exits.
.TP
\fB-?\fR, \fB--help\fR
Displays help, & then exits.
.SS "Selection"
.TP
\fB-M\fR \fIBytes\fR, \fB--maximumBytes=\fR\fIBytes\fR
Defines the maximum available space, in bytes; defaulting to the space available on a DVD, i.e. \fB4700000000\fR bytes.
.TP
\fB-m\fR \fIFloat\fR, \fB--minimumUsageRatio='\fR\fIFloat\fR\fB'\fR
Defines the minimum acceptable usage-ratio (in the closed unit-interval [0,1]); defaulting to '\fB0.99\fR', i.e. 99% of \fBmaximumBytes\fR.
.TP
\fB-z\fR[\fIBool\fR], \fB--includeEmpty\fR[\fB=\fR\fIBool\fR]
When "\fBTrue\fR", any empty file or directory, can be a member of all solutions.
.br
The default value, in the absence of this option, is "\fBFalse\fR",
but in the absence of only the boolean argument, "\fBTrue\fR" will be inferred.
.br
CAVEAT: for \fIn\fR such files or directories, a factor of \fI2^n\fR times more viable solutions exist, obscuring the minimal solutions on which they're based.
.SS Test
.TP
\fB-r\fR[\fIInt\fR], \fB--randomSeed\fR[\fB=\fR\fIInt\fR]
This option takes an optional integral argument with which to seed the unique random-number generator used for all random operations.
.br
In the absence of this option, the random-number generator will be seeded unpredictably from the operating-system,
but in the absence of only the integral argument, "\fB0\fR" will be inferred.
.br
CAVEAT: to be effective, this option must precede either "\fBtestPerformanceContinuous\fR" or "\fBtestPerformanceDiscrete\fR".
.TP
\fB--testPerformanceContinuous='(\fR\fIInteger\fR\fB,\fR \fBLogNormalDistribution\fR \fIlocation\fR \fIscale^2\fR\fB)'\fR
Measures the CPU-time required to find the best fit,
for the specified number of randomly generated virtual files,
the size of which conform to the specified continuous probability-distribution; & then exits.
.br
"\fIA Large-Scale Study of File-System Contents by John R. Douceur and William J. Bolosky\fR" concludes that for arbitrary file-types,
the frequency of file-sizes matches this distribution.
.TP
\fB--testPerformanceDiscrete='(\fR\fIInteger\fR\fB,\fR \fBPoissonDistribution\fR \fIlambda\fR\fB)'\fR
Measures the CPU-time required to find the best fit,
for the specified number of randomly generated virtual files,
the size of which conform to the specified discrete probability-distribution; & then exits.
.br
CAVEAT: the standard-deviation for this distribution is narrower than observed.
.SS File-paths
.TP
If \fIFile-path\fR is a single hyphen-minus ("\fB-\fR"), then the list of file-paths will be read from standard-input.
.SH EXIT-STATUS
\fB0\fR on success, & >\fB0\fR if an error occurs.
.SH EXAMPLES
.SS Example 1
Say we've a directory of audio-files, categorised by artist.
.IP
.B ls -p
.nf
	ArabStrap/ BobDylan/ JeffBuckley/ JohnMartyn/ JoniMitchell/ ReservoirDogsOST/
	RichardThompson/ SethLakeman/ SusheelaRaman/ TeddyThompson/ Vangelis/
.fi
.PP
We want to find which combinations can be stored on a CD without wasting inordinate amounts of space.
.IP
.B squeeze -M 700000000 -m 0.999 *
.nf
	699871313	["BobDylan","RichardThompson","SethLakeman","TeddyThompson"]
	699893320	["ArabStrap","BobDylan","JeffBuckley","JohnMartyn","SethLakeman","SusheelaRaman"]
	\fI699998310\fR	["ArabStrap","BobDylan","JoniMitchell","ReservoirDogsOST","SethLakeman","TeddyThompson","Vangelis"]
.fi
.PP 
Note that the proposed solutions don't split any of the directories, into their constituent files.
.PP
We can confirm the validity of the optimal result:
.IP
.B find ArabStrap BobDylan JoniMitchell ReservoirDogsOST SethLakeman TeddyThompson Vangelis -type f -print | perl -e 'use List::Util qw(sum); printf(qq(%d\\n), sum map { chomp; (stat)[7] } <>);'
.nf
	699998310
.fi
.PP
N.B.: "\fBdu\fR" will return a slightly larger size, since it includes the space required for directory-structures.
.SS Example 2
We can improve on that result if we're prepared to split some of the artist-specific directories into individual albums.
With the expectation of more solutions from which to select, we'll also raise the bar for what is acceptable.
.br
N.B.: from version \fB1.0.3.0\fR, the runtime can be passed a "\fB-N\fR" flag to request parallel execution on multiple cores,
which as a side-effect, alters the selection of sub-optimal solutions to return.
Whether this option actually results in reduced execution-time, depends on the file-size frequency-distribution & on the available hardware.
.IP
.B squeeze -M 700000000 -m 0.99999 ArabStrap BobDylan/* JeffBuckley JohnMartyn JoniMitchell ReservoirDogsOST RichardThompson SethLakeman SusheelaRaman TeddyThompson Vangelis +RTS -N
.nf
	699995815	["ArabStrap","BobDylan/BlondeOnBlonde","BobDylan/Highway61Revisited","BobDylan/ModernTimes","BobDylan/StreetLegal","BobDylan/SubterraneanHomesickBlues","JoniMitchell","RichardThompson","SusheelaRaman","TeddyThompson","Vangelis"]
	699998578	["BobDylan/Desire","BobDylan/Highway61Revisited","BobDylan/Infidels","BobDylan/StreetLegal","JeffBuckley","JohnMartyn","JoniMitchell","RichardThompson","SethLakeman","SusheelaRaman","TeddyThompson","Vangelis"]
	\fI699999112\fR	["BobDylan/BlondeOnBlonde","BobDylan/Desire","BobDylan/Highway61Revisited","BobDylan/Infidels","BobDylan/ModernTimes","JeffBuckley","JohnMartyn","ReservoirDogsOST","RichardThompson","SusheelaRaman","TeddyThompson","Vangelis"]
.fi
.SS Example 3
If we're prepared to add individual files from another artist:
.IP
.B squeeze -M 700000000 -m 0.9999999 ArabStrap BobDylan/* JeffBuckley JohnMartyn JoniMitchell ReservoirDogsOST $(find RichardThompson -type f) SethLakeman SusheelaRaman TeddyThompson Vangelis
.nf
	699999964	["ArabStrap","BobDylan/Desire","BobDylan/Infidels","BobDylan/ModernTimes","BobDylan/OhMercy","BobDylan/SubterraneanHomesickBlues","BobDylan/TimeOutOfMind","JeffBuckley","JohnMartyn","JoniMitchell","ReservoirDogsOST","RichardThompson/FrontParlourBallads/RichardThompson-06-MySoulMySoul.ogg","RichardThompson/RumorAndSigh/RichardThompson-12-MotherKnowsBest.ogg","RichardThompson/TheOldKitBag/RichardThompson-01-Gethsemane.ogg","RichardThompson/TheOldKitBag/RichardThompson-11-OutsideOfTheInside.ogg","SethLakeman","SusheelaRaman","TeddyThompson","Vangelis"]
	699999987	["BobDylan/BlondeOnBlonde","BobDylan/BloodOnTheTracks","BobDylan/Desire","BobDylan/ModernTimes","BobDylan/OhMercy","BobDylan/StreetLegal","BobDylan/SubterraneanHomesickBlues","BobDylan/TimeOutOfMind","JoniMitchell","ReservoirDogsOST","RichardThompson/FrontParlourBallads/RichardThompson-01-LetItBlow.ogg","RichardThompson/IWantToSeeTheBrightLightsTonight/RichardAndLindaThompson-13-TheCalvaryCross[Live-Bonus].ogg","RichardThompson/TheOldKitBag/RichardThompson-01-Gethsemane.ogg","RichardThompson/TheOldKitBag/RichardThompson-06-FirstBreath.ogg","RichardThompson/TheOldKitBag/RichardThompson-11-OutsideOfTheInside.ogg","SethLakeman","SusheelaRaman","TeddyThompson","Vangelis"]
	\fI700000000\fR	["ArabStrap","BobDylan/Desire","BobDylan/Highway61Revisited","BobDylan/Infidels","BobDylan/ModernTimes","BobDylan/StreetLegal","BobDylan/SubterraneanHomesickBlues","BobDylan/TimeOutOfMind","JeffBuckley","JoniMitchell","ReservoirDogsOST","RichardThompson/FrontParlourBallads/RichardThompson-01-LetItBlow.ogg","RichardThompson/RumorAndSigh/RichardThompson-08-BacklashLoveAffair.ogg","RichardThompson/RumorAndSigh/RichardThompson-12-MotherKnowsBest.ogg","RichardThompson/TheOldKitBag/RichardThompson-01-Gethsemane.ogg","RichardThompson/TheOldKitBag/RichardThompson-04-ALoveYouCantSurvive.ogg","RichardThompson/TheOldKitBag/RichardThompson-06-FirstBreath.ogg","SusheelaRaman","TeddyThompson","Vangelis"]
.fi
.PP
.B ^C
.PP
The exact match isn't unexpected, given the 2^71 possible combinations.
The process was terminated after this solution was found, though where time permits, one may choose to wait for alternative exact matches.
.SH AUTHOR
Written by Dr. Alistair Ward.
.SH BUGS
.SS "REPORTING BUGS"
Report bugs to <\fBsqueeze@functionalley.com\fR>.
.SH COPYRIGHT
Copyright \(co 2010-2015 Dr. Alistair Ward
.PP
This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with this program. If not, see <\fBhttps://www.gnu.org/licenses/\fR>.
.SH "SEE ALSO"
.IP \(bu
Home-page: <\fBhttps://functionalley.com/Squeeze/squeeze.html\fR>.
.IP \(bu
<\fBhttps://hackage.haskell.org/package/squeeze\fR>.
.IP \(bu
<\fBhttps://github.com/functionalley/Squeeze\fR>.
.IP \(bu
Source-documentation is generated by "\fBHaddock\fR", & is available in the distribution.
.IP \(bu
<\fBhttps://www.haskell.org/haddock/\fR>.
.IP \(bu
<\fBhttps://en.wikipedia.org/wiki/Log-normal_distribution\fR>.
.IP \(bu
<\fBhttps://en.wikipedia.org/wiki/Poisson_distribution\fR>.


changelog.markdown
# 2010-11-18 Dr. Alistair Ward <squeeze@functionalley.com>

## 1.0.0.0
* First version of the package.

## 1.0.0.1
* Created sub-directory "**src/Squeeze/**" & then modified module-names accordingly.
* Pacified **hlint**.
* Added a makefile.

## 1.0.0.2
* Tested with **ghc-7.0.1**.
* Added *.spec*-file.

## 1.0.1.0
* Improved the basic algorithm, by permitting the file-list to be bisected; each file-combination resulting from the LHS, is then concatenated with each of those resulting from the recursively bisected RHS.
* Used module "**Control.Parallel.Strategies**" to parallelize the concatenation of each combination from the LH-list, with each of those from the RH-list.
* Added default values of command-line flags, to usage-message.
* Used new module "**ToolShed.Package**".
(1.0.1.1)
* Replaced `Data.List.subsequences` with a tailored version which terminates combinations when they've grown too large, providing a better and more predictable performance-gain than recursive list-bisection.
* Reduced recursive list-bisection to a single bisection, for the sole purpose of facilitating parallelisation.
* Added module "**Squeeze.QC**" to define **Test.QuickCheck**-tests and command-line option "**--verbose**", to govern level of output.
* Partitioned into new modules; "**Squeeze.File**", "**Squeeze.FileSizeBounds**" and "**Squeeze.FileCombination**".

## 1.0.1.2
* Fixed typo in CPP-statement in module "**Squeeze.QC**".
* Fixed negated reference to unparenthesised CPP-macro "**MIN_VERSION**".
* Protected reference in module "**Main**", to `Test.QuickCheck.verboseCheck`, using CPP-statement.

## 1.0.1.3
* Amended *.spec*-file to define RPM-package, to install under directory "**/usr**", rather than "**/usr/local**".
* Renamed package from "**Squeeze**" to "**squeeze**", for compatibility with Debian's *.deb*-format.
* Added directory-structure required to build *.deb*-package.

## 1.0.1.4
* Corrected command-line processing of the special file "**-**", which reads standard-input as requested, but disregarded any other command-line options.
* Refined man-page to be an "**extra-source-file**" in the *.cabal*-file and the *.spec*-file.
* Ported to compile against alternative integral definitions of `Squeeze.File.FileSize`.
* Added manually controlled flag "**llvm**" to the *.cabal*-file.
* Replaced references to the module "**ToolShed.Package**" with "**Distribution.Package**".
* Reacted to the creation of module "**ToolShed.Defaultable**".
* Uploaded to [Hackage](https://hackage.haskell.org/package/squeeze).

## 1.0.2.0
* Reimplemented `Squeeze.File.aggregateSize` using `foldr` rather than `foldl`.
* Removed use of `Control.Parallel.Strategies` from **Squeeze.Squeeze**, which ruined performance on one core and had no obvious benefit on two.
* Changed the definition of `Squeeze.FileCombinations.FileCombinations` from a `Pair` to a record-style data-type with selective strictness-annotation.
* Renamed `Squeeze.FileCombination.concatenate` to `(Squeeze.FileCombination.<+>)` and `Squeeze.FileCombination.prepend` to `(Squeeze.FileCombination.+>)`.
* Used new module "**ToolShed.TimeAction**", to permit measurement of the CPU-time required to find the best fit.
* Added the number, mean file-size and standard-deviation, to the output, when flag "**--verbose**" is specified; and as a side-effect, created a dependency on the package "**factory**".
* Replaced module "**Squeeze.FileSizeBounds**" with **Factory.Data.Interval**, requiring use of **toolshed-0.12.0.0**.
* Made the data-type `CommandOptions.CommandOptions` polymorphic in terms of the *ratio*-type fields.
* Inserted new layers **Data** and **Test** into the module-hierarchy, and relocated existing modules accordingly.
* Added module "**Squeeze.Test.Performance**" to test with randomly generated virtual files.

## 1.0.2.1
* Added check for zero file-names when reading from stdin.
* Corrected the syntax required when specifying ratios, in the man-page.

## 1.0.2.2
* Renamed `Squeeze.Data.File.order` to `Squeeze.Data.File.orderBySize`.
* Used **Paths_squeeze.version** in **Main**, rather than hard-coding it.
* Added `Squeeze.Data.CommandOptions.getErrors` to permit better error-diagnosis.
* Reacted to new module-hierarchy and addition of method `ToolShed.SelfValidate.getErrors`, in toolshed-0.13.0.0.
* Replaced use of `Control.Exception.bracket` with `System.IO.withFile`, in `Squeeze.Data.File.findSize`.
* Replaced `System` with `System.Environment` and `System.Exit`.
* Removed dependency on **haskell98**.

## 1.0.2.3
* Added class `Show` to context for `Squeeze.Test.Performance.run`, for migration to **ghc-7.4.1**.

## 1.0.2.4
* Added details to any failure to parse the command-line arguments.
* Defined package's name using program's name & enhanced usage-message, in **Main.hs**.
* Re-implemented logging using `Control.Monad.Writer.Writer` & renamed `Squeeze.Squeeze.findBestFitM` to `Squeeze.Squeeze.findBestFitWriter`.
* Implemented **verbose** functionality in **Squeeze.Test.Performance**.
* Added command-line flag "**includeEmpty**", to allow one to include empty files & directories in any solution.

## 1.0.3.0
* Changed **Main** to replace calls to `error` from inside the IO-monad, with `Control.Monad.fail`.
* Changed **Main**; by adding a command-line option to define the seed for the random-generator used for performance-testing; to permit **includeEmpty** to take an optional Boolean argument, rather than a mandatory one; to read `Squeeze.Data.CommandOptions.getMinimumUsageRatio` as a `Double`, rather than a `Rational`.
* Renamed the accessors in **Squeeze.Data.CommandOptions**, & removed `bisectionRatio` & associated code.
* Added **Squeeze.Control.Concurrent.DivideAndConquer**, called from **Main** to utilise multiple CPU-cores where available.
* Trapped command-line arguments to which garbage has been appended.
* Changed **Data.CommandOptions** to use `Distribution.Verbosity.Verbosity`.

## 1.0.4.0

src-lib/Squeeze/Squeeze.hs
{-
	Copyright (C) 2010-2016 Dr. Alistair Ward

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
-}
{- |
 [@AUTHOR@]	Dr. Alistair Ward

 [@DESCRIPTION@]

	* Returns combinations of the specified files, which fit into the available space, without wasting more than the specified ratio.

	* Any directory-names are treated as atomic units, rather than individual files.

	* Because of the exponential growth of possible combinations,
	an /exact/ match for the available space is frequently found with a surprisingly small set of files.
-}

module Squeeze.Squeeze(
-- * Functions
	findCombinations,
	findBestFit,
	distributeAndFindBestFit,
	partitionEmptyFilesAndDistributeAndFindBestFit
) where

import qualified	Control.Arrow
import			Control.Arrow((&&&))
import qualified	Control.Concurrent
import qualified	Control.Monad
import qualified	Data.List
import qualified	Factory.Data.Interval
import qualified	Squeeze.Control.Concurrent.DivideAndConquer	as Control.Concurrent.DivideAndConquer
import qualified	Squeeze.Data.CommandOptions			as Data.CommandOptions
import qualified	Squeeze.Data.File				as Data.File
import qualified	Squeeze.Data.FileCombination			as Data.FileCombination
import qualified	System.IO

{- |
	* Checks that the total aggregate 'Data.File.FileSize', meets or exceeds 'minimumBytes'.

	* Drops excessively large files, assuming that the file-list has been sorted by size, largest first.

	* Generates up to @2^n@ combinations of the @n@ specified files; the algorithm is similar to 'Data.List.subsequences', except that unproductive lines are immediately terminated.
	This is the performance bottle-neck, and though there may be simpler and faster algorithms, the key attribute is that it operates in constant space.

	* CAVEAT: assumes files have been sorted by decreasing size.

	* The algorithm is stable, in that it maintains the specified file-order within each combination;
	though the order in which the combinations are concatenated is rather arbitrary.
-}
findCombinations
	:: Factory.Data.Interval.Interval Data.File.FileSize	-- ^ The closed interval of acceptible aggregate size, for file-combinations.
	-> [Data.File.FileSizeAndPath]				-- ^ The list of file-names & sizes, ordered by decreasing size.
	-> [Data.FileCombination.FileCombination]		-- ^ The resulting unordered list of suitable file-combinations.
findCombinations (minimumCombinationSize, maximumCombinationSize)	= filter (
	Data.FileCombination.hasSizeBy (>= minimumCombinationSize)
 ) . (
	Data.FileCombination.nullFileCombination :
 ) . nonEmptyCombinations minimumCombinationSize . uncurry zip . (
	id &&& Data.File.accumulateSize			-- Associate the list of possible files with its accumulating size.
 ) . dropWhile (
	Data.File.hasSizeBy (> maximumCombinationSize)	-- Remove files which individually exceed the maximum permissible; assuming they've been sorted by decreasing size.
 ) where
	nonEmptyCombinations :: Data.File.FileSize -> [(Data.File.FileSizeAndPath, Data.File.FileSize)] -> [Data.FileCombination.FileCombination]
	nonEmptyCombinations _ []	= []
	nonEmptyCombinations minimumBytes ((fileSizeAndPath, aggregateSize) : remainder)
		| aggregateSize < minimumBytes	= []	-- Even if all the files are selected, the minimum-size criterion won't be satisfied.
		| otherwise			= Data.FileCombination.singleton fileSizeAndPath : foldr binaryChoice [] (
			nonEmptyCombinations (minimumBytes - Data.File.getSize fileSizeAndPath) remainder	-- Recurse.
		)
		where
			binaryChoice :: Data.FileCombination.FileCombination -> [Data.FileCombination.FileCombination] -> [Data.FileCombination.FileCombination]
			binaryChoice combinationExcluding combinations
				| Data.FileCombination.hasSizeBy (> maximumCombinationSize) combinationIncluding	= combinations'
				| otherwise										= combinationIncluding : combinations'
				where
					combinationIncluding :: Data.FileCombination.FileCombination
					combinationIncluding	= Data.FileCombination.prepend fileSizeAndPath combinationExcluding

					combinations'	= combinationExcluding : combinations

-- | Orders the files by decreasing size, calls 'findCombinations', calls 'Data.FileCombination.risingFilter' to select progressively better solutions.
findBestFit
	:: Factory.Data.Interval.Interval Data.File.FileSize	-- ^ The closed interval of acceptible sizes for file-combinations.
	-> [Data.File.FileSizeAndPath]				-- ^ The input list of file-names & sizes.
	-> [Data.FileCombination.FileCombination]		-- ^ A reduced list of increasingly suitable file-combinations.
findBestFit solutionSizeBounds	= Data.FileCombination.risingFilter (Factory.Data.Interval.getMinBound solutionSizeBounds) . findCombinations solutionSizeBounds . Data.File.orderByDecreasingSize {-which makes findCombinations faster-}

{- |
	* Recursively bisects the task, distributing the sub-tasks to 'findBestFit', to utilise the available CPU-cores.

	* The task is bisected by removing the smallest file, then solving the remaining problem for two independent cases; that the selected file is excluded or the selected file is included, in the final solution.
	Selecting the smallest file rather than the largest, seems to balance the load of the sub-tasks more evenly.
	CAVEAT: no account has been taken of the possibility that the smallest file has size zero, which makes the sub-tasks identical.

	* Recombines the part solutions to finds the single monotonically increasing list of file-combinations matching the original criteria.

	* CAVEAT: whilst the ultimate solution is similar, regardless of the specified number of CPU-cores available, the path leading to it typically differs.
-}
distributeAndFindBestFit
	:: RealFrac ratio
	=> Data.CommandOptions.CommandOptions ratio
	-> [Data.File.FileSizeAndPath]	-- ^ The unordered list of files & sizes.
	-> IO [Data.FileCombination.FileCombination]
distributeAndFindBestFit commandOptions fileSizeAndPathList	= let
	slave _ _ []						= return {-to IO-monad-} [Data.FileCombination.nullFileCombination]
	slave 1	commandOptions' increasingFileSizeAndPathList	= let
		solutionSizeBounds	= Data.CommandOptions.solutionSizeBounds commandOptions'
	 in do
		Control.Monad.when (Data.CommandOptions.getVerbosity commandOptions' == maxBound) . System.IO.hPutStrLn System.IO.stderr $ "INFO: acceptable file-size interval " ++ show solutionSizeBounds ++ " bytes."

		return {-to IO-monad-} $ findBestFit solutionSizeBounds increasingFileSizeAndPathList	-- Delegate the task to the single-threaded algorithm.
	slave numCapabilities' commandOptions' (selectedFileSizeAndPath {-the smallest-} : remainingFileSizeAndPaths)	= let
		recurse	= slave $ numCapabilities' `div` 2	-- Partially apply.
	 in do
		Control.Monad.when (Data.CommandOptions.getVerbosity commandOptions' == maxBound) . System.IO.hPutStrLn System.IO.stderr $ "INFO: " ++ show numCapabilities' ++ " CPU-cores => bisecting task into those, with & without " ++ show selectedFileSizeAndPath ++ "."

		fileCombinationsExcludingSelected	<- recurse commandOptions' {-i.e the original space-} remainingFileSizeAndPaths	-- This is about half the total task.

		let
			commandOptions''	= Data.CommandOptions.subtractFile (Data.File.getSize selectedFileSizeAndPath) commandOptions' -- Reduce the requirements by the size of the selected file.

		if Data.CommandOptions.getMaximumBytes commandOptions'' < 0
			then return {-to IO-monad-} fileCombinationsExcludingSelected	-- The selected file won't fit.
			else map (
				Data.FileCombination.prepend selectedFileSizeAndPath	-- Prepend the selected file to all file-combinations.
			) `fmap` recurse commandOptions'' remainingFileSizeAndPaths >>= Control.Concurrent.DivideAndConquer.divideAndConquer Data.FileCombination.risingMergeByAggregateFileSize fileCombinationsExcludingSelected	-- Merge the part-solutions.
 in Control.Concurrent.getNumCapabilities >>= (\numCapabilities -> slave numCapabilities commandOptions $ Data.File.orderByIncreasingSize fileSizeAndPathList)

{- |
	* Neither 'distributeAndFindBestFit' nor 'findBestFit' distinguish between empty & non-empty files,
	but empty files cause significant inefficiency in the former (where the same calculation is performed multiple times)
	& could be treated much more efficiently in the latter (since they're potentially a member of any other solution).

	* This function side-lines empty files, delegates the remaining problem to 'distributeAndFindBestFit' (& consequently 'findBestFit'),
	then prepends combinations of empty files to the resulting combinations of non-empty files.
-}
partitionEmptyFilesAndDistributeAndFindBestFit
	:: RealFrac ratio
	=> Data.CommandOptions.CommandOptions ratio
	-> [Data.File.FileSizeAndPath]	-- ^ The unordered list of files & sizes.
	-> IO [Data.FileCombination.FileCombination]
partitionEmptyFilesAndDistributeAndFindBestFit commandOptions fileSizeAndPathList
	| Data.CommandOptions.getIncludeEmpty commandOptions	= do
		printStatistics fileSizeAndPathList

		concatMap (
			\fileCombination -> map (
				\emptyFilePathCombination	-> fileCombination {
					Data.FileCombination.getFilePathList	= emptyFilePathCombination ++ Data.FileCombination.getFilePathList fileCombination
				}
			) $ Data.List.subsequences {-find all combinations-} emptyFiles
		 ) `fmap` nonEmptyFileCombinations
	| otherwise {-exclude empty files-}			= do
		Control.Monad.unless (Data.CommandOptions.getVerbosity commandOptions == minBound || null emptyFiles) . System.IO.hPutStrLn System.IO.stderr $ "WARNING: rejecting empty files; " ++ show emptyFiles ++ "."

		Control.Monad.when (null nonEmptyFilePathAndSizeList) $ error "there are zero non-empty files."

		printStatistics nonEmptyFilePathAndSizeList

		nonEmptyFileCombinations
	where
		printStatistics l				= Control.Monad.unless (Data.CommandOptions.getVerbosity commandOptions == minBound || null l) . System.IO.hPutStrLn System.IO.stderr $ "INFO: file-(count, aggregate size, mean, standard-deviation); " ++ show (Data.File.getFileSizeStatistics l :: (Int, Data.File.FileSize, Float, Float)) ++ "."
		(emptyFiles, nonEmptyFilePathAndSizeList)	= Control.Arrow.first (map Data.File.getPath) $ Data.List.partition (Data.File.hasSizeBy (== 0)) fileSizeAndPathList
		nonEmptyFileCombinations			= distributeAndFindBestFit commandOptions nonEmptyFilePathAndSizeList	-- Delegate.

